uniffi_bindgen/src/bindings/kotlin/templates/TopLevelFunctionTemplate.kt

```diff
@@ -1,3 +1,68 @@
+{%- if func.is_async() %}
+{%- match func.throws_type() -%}
+{%- when Some with (throwable) %}
+@Throws({{ throwable|type_name }}::class)
+{%- else -%}
+{%- endmatch %}
+{%- match func.return_type() -%}
+{%- when Some with (return_type) %}
+
+suspend fun {{ func.name()|fn_name }}({%- call kt::arg_list_decl(func) -%}) = suspendCoroutine<{{ return_type|type_name }}> { continuation ->
+    class Waker: RustFutureWaker {
+        override fun callback(envCStructure: RustFutureWakerEnvironmentCStructure?) {
+            @OptIn(DelicateCoroutinesApi::class)
+            GlobalScope.launch {
```

Comment:
Why is it needed to run this in GlobalScope?

Reply:
Fixed in https://github.com/mozilla/uniffi-rs/pull/1409/commits/67d78756cb1f43c1c8b100bb04cdcd0e73afe23c.

Reply:
And the new approach works? As far as I know `coroutineScope { }` is only active within its bounds (between `{` and `}`). So later `launch` is called on a finished `CoroutineScope`...

Reply:
It works yes. The object is not dropped so I think as long as `launchˋ is called during its lifetime, everything should be fine?

Reply:
I think it is a violation of structured concurrency. Why is a `launch` needed at all?

Reply:
The code inside `coroutineScope` is blocked by `suspendCoroutine`. It's literally waiting/suspended. When `suspendCoroutine` ends, then only at that point, `coroutineScope` ends. The generated code looks like this:

```kotlin
    val result: …

    coroutineScope {
        result = suspendCoroutine<…> { continuation ->
            // do something with `continuation` and `this` (`CoroutineScope`)
        }
        // point A
    }

    return result
```

We reach `point A` (end of `coroutineScope`) only when `suspendCoroutine` returns.

To answer to the question: Why is a `launch` needed at all`, please refer to the documentation: https://github.com/Hywan/uniffi-rs/blob/48ed0bd5095c48f1c998c9c25d9c2e83b4e73eef/uniffi/src/ffi/rustfuture.rs#L163-L206

Reply:
You are right, sorry. Still don't understand, why `launch` is needed, but as long as it works it's not that important to me.

Reply:
The `waker` must not `poll` the Rust `Future` directly, but it must schedule the foreign language's executor to `poll` the Rust `Future` when possible. So `waker` must return immediately after having scheduled the `poll`, hence `launch`.

Reply:
Now it's clear, thank you :)